spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java

```diff
@@ -324,12 +326,37 @@ else if (!(headerName.equals(JSON_TYPES)) && matchesForInbound(headerName)) {
 					populateJsonValueHeader(header, requestedType, headers);
 				}
 				else {
-					headers.put(headerName, headerValueToAddIn(header));
+					handleHeader(headerName, header, headers);
 				}
 			}
 		});
 	}
 
+	/**
+	 * Handle non-reserved headers in {@link DefaultKafkaHeaderMapper}.
+	 * @param headerName the header name.
+	 * @param header the header instance.
+	 * @param headers the target headers.
+	 * @since 4.0.0
+	 */
+	protected void handleHeader(String headerName, Header header, final Map<String, Object> headers) {
+		if (!this.doesMatchMultiValueHeader(headerName)) {
+			headers.put(headerName, headerValueToAddIn(header));
+		}
+		else {
+			Object values = headers.getOrDefault(headerName, new ArrayList<>());
+			if (values instanceof List) {
```

Comment:
I'm not fully sure in this logic.
Right now you mean: if there is a header with requested name and it is not a list, then we override its value.
Is it really expected?
Why we just cannot always override withe list since this is exactly what is asked by the respective pattern?
Otherwise I feel like we cannot override since the value is already there.